src/test/result_tests.cpp

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <util/result.h>
+
+#include <boost/test/unit_test.hpp>
+
+BOOST_AUTO_TEST_SUITE(result_tests)
+
+util::Result<void> VoidSuccessFn()
+{
+    return {};
+}
+
+util::Result<void> VoidFailFn()
+{
+    return {util::Error{}, Untranslated("void fail")};
+}
+
+util::Result<int> IntSuccessFn(int ret)
+{
+    return {ret};
+}
+
+util::Result<int> IntFailFn()
+{
+    return {util::Error{}, Untranslated("int fail")};
+}
+
+enum FnStatus { SUCCESS, ERR1, ERR2 };
+
+util::Result<FnStatus> StatusSuccessFn(FnStatus ret)
+{
+    return {ret};
+}
+
+util::Result<FnStatus> StatusFailFn(FnStatus ret)
+{
+    return {util::Error{}, Untranslated("status fail"), ret};
+}
+
+util::Result<int> ChainedFailFn(FnStatus arg, int ret)
+{
+    return {util::ErrorChain{}, Untranslated("chained fail"), StatusFailFn(arg), ret};
```

Comment:
It would be possible and not very difficult to change syntax from

```c++
return {util::Error{}, "error string", error_value};
``` 

to

```c++
return util::Error("error string", error_value);
```

for both Error and ErrorChain with some template magic. It just made the implementation a little uglier so I decided to drop it.